[18.0][MIG] account_invoice_refund_line_selection

imlopes opened the pull request:
Main changes :

The field `refund_method` do not exists anymore.
So I added only a boolean that allow the user to create a refund for +one move line.

![image](https://github.com/user-attachments/assets/90c9dac0-5096-4127-a829-ed0d23833c82)

Commit:
[ADD] account_invoice_refund_line

Commit:
[ADD] account_invoice_refund_line

OCA-git-bot committed:
[ADD] icon.png

OCA-git-bot committed:
[ADD] icon.png

Commit:
[12.0][MIG] account_invoice_refund_line_selection

Commit:
[12.0][MIG] account_invoice_refund_line_selection

etobella committed:
[IMP] account_invoice_refund_line_selection: black, isort, prettier

etobella committed:
[IMP] account_invoice_refund_line_selection: black, isort, prettier

etobella committed:
[MIG] account_invoice_refund_line_selection: Migration to 13.0

etobella committed:
[MIG] account_invoice_refund_line_selection: Migration to 13.0

etobella committed:
[IMP] account_invoice_refund_line_selection: Add view for search

etobella committed:
[IMP] account_invoice_refund_line_selection: Add view for search

olgamarcocb committed:
[IMP] account_invoice_refund_line_selection: black, isort, prettier

olgamarcocb committed:
[IMP] account_invoice_refund_line_selection: black, isort, prettier

olgamarcocb committed:
[MIG] account_invoice_refund_line_selection: Migration to 14.0

olgamarcocb committed:
[MIG] account_invoice_refund_line_selection: Migration to 14.0

oca-travis committed:
[UPD] Update account_invoice_refund_line_selection.pot

oca-travis committed:
[UPD] Update account_invoice_refund_line_selection.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

ofonseca-pyming committed:
Added translation using Weblate (Catalan)

ofonseca-pyming committed:
Added translation using Weblate (Catalan)

kittiu committed:
account_invoice_refund_line_selection, error singleton

kittiu committed:
account_invoice_refund_line_selection, error singleton

bosd committed:
Added translation using Weblate (Dutch)

bosd committed:
Added translation using Weblate (Dutch)

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 81.8% (9 of 11 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_refund_line_selection/nl/

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 81.8% (9 of 11 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_refund_line_selection/nl/

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 90.9% (10 of 11 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_refund_line_selection/nl/

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 90.9% (10 of 11 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_refund_line_selection/nl/

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.1

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.1

willytal committed:
[FIX] account_invoice_refund_line_selection: fix tests

TT32113

willytal committed:
[FIX] account_invoice_refund_line_selection: fix tests

TT32113

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.2

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.2

jappi00 committed:
[FIX] account_invoice_refund_line_selection: Business fields were missing

jappi00 committed:
[FIX] account_invoice_refund_line_selection: Business fields were missing

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.3

OCA-git-bot committed:
account_invoice_refund_line_selection 14.0.1.0.3

ChrisOForgeFlow committed:
[15.0][MIG] account_invoice_refund_line_selection

ChrisOForgeFlow committed:
[15.0][MIG] account_invoice_refund_line_selection

RodrigoBM committed:
[MIG] account_invoice_refund_line_selection: Migration to 16.0

RodrigoBM committed:
[MIG] account_invoice_refund_line_selection: Migration to 16.0

Commit:
[UPD] Update account_invoice_refund_line_selection.pot

Commit:
[UPD] Update account_invoice_refund_line_selection.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/es/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/es/

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 12.5% (1 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 12.5% (1 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/it/

rousseldenis committed:
Added translation using Weblate (French)

rousseldenis committed:
Added translation using Weblate (French)

rousseldenis committed:
Translated using Weblate (French)

Currently translated at 62.5% (5 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/fr/

rousseldenis committed:
Translated using Weblate (French)

Currently translated at 62.5% (5 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_refund_line_selection
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_refund_line_selection/fr/

imlopes committed:
[IMP] account_invoice_refund_line_selection: pre-commit auto fixes

imlopes committed:
[IMP] account_invoice_refund_line_selection: pre-commit auto fixes

imlopes committed:
[18.0][MIG] account_invoice_refund_line_selection

imlopes committed:
[18.0][MIG] account_invoice_refund_line_selection

ivantodorovich reviewed (approved):
technically LG

yankinmax commented on account_invoice_refund_line_selection/README.rst:
```diff
@@ -0,0 +1,84 @@
+Usage
+=====
+
+In a open validated invoice, click on 'Add Credit Note' to open the
```
```
Click the Add Credit Note button to launch the refunding wizard in an open, validated invoice. Then, choose the Refund specific lines option and select the invoice lines you want to refund.

yankinmax commented on account_invoice_refund_line_selection/tests/test_refund_line.py:
```diff
@@ -0,0 +1,138 @@
+    @classmethod
+    def setUpClass(cls, chart_template_ref=None):
+        super().setUpClass()
+        cls.env = cls.env(
```
I think you don't need to mute mail-related context, because it's already done in `BaseCommon`
See: `class AccountTestInvoicingCommon(ProductCommon)` -> `class ProductCommon(BaseCommon, UomCommon)`

yankinmax commented on account_invoice_refund_line_selection/tests/test_refund_line.py:
```diff
@@ -0,0 +1,138 @@
+        cls.out_invoice._post()
+
+    def test_partial_refund_in_invoice(self):
+        reversal = (
```
I would test with a Form, because there usually exist some on-changes on the wizards.

yankinmax commented on account_invoice_refund_line_selection/wizards/account_move_reversal.py:
```diff
@@ -0,0 +1,53 @@
+        if active_id:
+            inv = self.env["account.move"].browse(active_id)
+            rec.update(
+                {"selectable_invoice_lines_ids": [(6, 0, inv.invoice_line_ids.ids)]}
```
```suggestion
                {"selectable_invoice_lines_ids": [Command.set(inv.invoice_line_ids.ids)]}
```

yankinmax commented on account_invoice_refund_line_selection/wizards/account_move_reversal.py:
```diff
@@ -0,0 +1,53 @@
+            lines = []
+            for line in reversal_inv.line_ids:
+                dict_line = line._convert_to_write(line._cache)
+                lines.append((0, 0, dict_line))
```
```suggestion
                lines.append(Command.create(dict_line_values))
```

yankinmax commented on account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml:
```diff
@@ -0,0 +1,42 @@
+                        <field name="selectable_invoice_lines_ids" invisible="1" />
+                        <field
+                            name="line_ids"
+                            context="{'tree_view_ref': 'account.view_move_line_tree_grouped_sales_purchases'}"
```
```suggestion
                            context="{'list_view_ref': 'account.view_move_line_tree_grouped_sales_purchases'}"
```

yankinmax commented on account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml:
```diff
@@ -0,0 +1,42 @@
+                            name="line_ids"
+                            context="{'tree_view_ref': 'account.view_move_line_tree_grouped_sales_purchases'}"
+                        >
+                            <list no_open="True">
```
can we add `readonly="1"` on a list?

imlopes replied:
Yes (at least, we have some samples on odoo code) : 

https://github.com/odoo/enterprise/blob/71e0092f6cef470347426eaabb6bb1cd4babce00/l10n_au_hr_payroll/views/hr_structure_type_views.xml#L12-L18

https://github.com/camptocamp/odoo/blob/233f03da5ae9c269618e6e28c775a123c0afca79/addons/hr/views/hr_employee_views.xml#L262-L266

Ricardoalso commented on account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml:
```diff
@@ -0,0 +1,42 @@
+            <xpath expr="//footer" position="before">
+                <notebook invisible="not refund_lines">
+                    <page string="Invoice Lines">
+                        <field name="selectable_invoice_lines_ids" invisible="1" />
```
```suggestion
```

Normally not necessary according to https://github.com/odoo/odoo/pull/137031

grindtildeath replied:
@imlopes Can you please attend this comment? :pray:

Ricardoalso reviewed (approved):
Small comment otherwise LGTM

leemannd commented:
Hello @OCA/accounting-maintainers , a review and an eventual merge is welcome

leemannd commented on account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml:
```diff
@@ -0,0 +1,42 @@
+            <xpath expr="//footer" position="before">
+                <notebook invisible="not refund_lines">
+                    <page string="Invoice Lines">
+                        <field name="selectable_invoice_lines_ids" invisible="1" />
```
```suggestion
```

imlopes committed:
account_invoice_refund_line_selection: Corrections after review

Update account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml
Co-authored-by: Denis Leemann <denis.leemann@camptocamp.com>

Update account_invoice_refund_line_selection/tests/test_refund_line.py
Co-authored-by: Sergio Corato <sergiocorato@users.noreply.github.com>

imlopes committed:
Update account_invoice_refund_line_selection/wizards/account_move_reversal_view.xml

Co-authored-by: Denis Leemann <denis.leemann@camptocamp.com>

dreispt commented:
All comments seem to have been attended.
A rebase might resolve the build issues.
If there are no objections I might attempt a merge soon.

dreispt commented:
/ocabot migration account_invoice_refund_line_selection 
/ocabot merge nobump

dreispt commented:
/ocabot migration account_invoice_refund_line_selection 
/ocabot merge nobump

dreispt commented:
/ocabot rebase

sergiocorato commented:
/ocabot merge nobump

sergiocorato commented:
/ocabot rebase

sergiocorato commented on account_invoice_refund_line_selection/tests/test_refund_line.py:
```diff
@@ -0,0 +1,154 @@
+        )
+        wiz_reversal_form.journal_id = self.out_invoice.journal_id
+        wiz_reversal_form.refund_lines = True
+        wiz_reversal_form.line_ids = line
```
```suggestion
        wiz_reversal_form.line_ids = line
        # check if module account_invoice_refund_reason is installed
        if hasattr(self.env["account.move"], "reason_id"):
            wiz_reversal_form.reason_id = self.env["account.move.refund.reason"].search(
                [], limit=1
            )
```

sergiocorato reviewed (changes requested):
@imlopes Should be resolved with this fix.

grindtildeath commented:
/ocabot merge nobump
